### ISSUE TITLE
Batch Inference For Alpaca

### ISSUE THREAD

**Issue by Yijia-Xiao** [facebookresearch/llama-recipes#163]
### System Info

PyTorch version: 2.0.1+cu118
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.2 LTS (x86_64)
GCC version: (Ubuntu 11.3.0-1ubuntu1~22.04.1) 11.3.0
Clang version: 14.0.0-1ubuntu1
CMake version: version 3.27.1
Libc version: glibc-2.35

Python version: 3.10.12 (main, Jul  5 2023, 18:54:27) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.15.0-75-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.0.140
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA A100-SXM4-80GB
GPU 1: NVIDIA A100-SXM4-80GB
GPU 2: NVIDIA A100-SXM4-80GB
GPU 3: NVIDIA A100-SXM4-80GB

Nvidia driver version: 530.30.02
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Address sizes:                   48 bits physical, 48 bits virtual
Byte Order:                      Little Endian
CPU(s):                          128
On-line CPU(s) list:             0-127
Vendor ID:                       AuthenticAMD
Model name:                      AMD EPYC 7543 32-Core Processor
CPU family:                      25
Model:                           1
Thread(s) per core:              2
Core(s) per socket:              32
Socket(s):                       2
Stepping:                        1
BogoMIPS:                        5589.57
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 invpcid_single hw_pstate ssbd mba ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 invpcid cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr rdpru wbnoinvd amd_ppin arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold v_vmsave_vmload vgif v_spec_ctrl umip pku ospke vaes vpclmulqdq rdpid overflow_recov succor smca
Virtualization:                  AMD-V
L1d cache:                       2 MiB (64 instances)
L1i cache:                       2 MiB (64 instances)
L2 cache:                        32 MiB (64 instances)
L3 cache:                        512 MiB (16 instances)
NUMA node(s):                    8
NUMA node0 CPU(s):               0-7,64-71
NUMA node1 CPU(s):               8-15,72-79
NUMA node2 CPU(s):               16-23,80-87
NUMA node3 CPU(s):               24-31,88-95
NUMA node4 CPU(s):               32-39,96-103
NUMA node5 CPU(s):               40-47,104-111
NUMA node6 CPU(s):               48-55,112-119
NUMA node7 CPU(s):               56-63,120-127
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Not affected
Vulnerability Retbleed:          Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Retpolines, IBPB conditional, IBRS_FW, STIBP always-on, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected

Versions of relevant libraries:
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.25.2
[pip3] torch==2.0.1+cu118
[pip3] torch-max-mem==0.0.5
[pip3] torch-ppr==0.0.8
[conda] torch                     2.0.1+cu118              pypi_0    pypi

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### 🐛 Describe the bug

I feed two samples into the user_prompt (i.e. `user_prompt=[prompt1_str, prompt2_str]`) in https://github.com/facebookresearch/llama-recipes/blob/main/inference/inference.py#L44. However, the inference result of `prompt1_str` differs from the result I get when I only feed one example. The `padding` parameter is set to `True` (https://github.com/facebookresearch/llama-recipes/blob/main/inference/inference.py#L102).

```
The response from the individual prompt feed: No, clasp knife spasticity is not typically a presentation of upper motor neuron lesions.
The response from 2 prompts (batch forward): leukoencephalitis spasticity is commonly due to UMN lesions and is a common presentation of spasticity.
```

The inference README says it supports batch forward,  I wonder if any other parameters should be set for the example scripts? For `alpaca` based fine-tuned models.

Thanks.

### Error logs

`[Expected output]` The response from the individual prompt feed: No, clasp knife spasticity is not typically a presentation of upper motor neuron lesions.
`[Problematic output from batched forward]` The response from 2 prompts (batch forward): leukoencephalitis spasticity is commonly due to UMN lesions and is a common presentation of spasticity.

### Expected behavior

Batched forward should have same output as sample-by-sample forward.

**Comment by Yijia-Xiao** [comment]
Further details: for sample-by-sample inference, if the `max_length` of the tokenizer is set to e.g. 1024, then the generated texts are also problematic.

The issue could be caused by `padding`. Some parameters in the inference script should be modified/updated.

**Comment by HaoyuanOuyang** [comment]
Have you solved this bug？
